src/printer/tvmscript_unified_printer.h

```diff
@@ -0,0 +1,656 @@
+#include <utility>
+
+#include "tvm/relay/expr.h"
+#include "tvm/runtime/container/array.h"
+#include "tvm/runtime/object.h"
+
+namespace tvm {
+namespace script {
+
+// Code Doc
+
+class DocNode : public Object {
+ public:
+  ObjectRef origin_ir_node;
+
+  DocNode() = default;
+  virtual ~DocNode() = default;
+
+  static constexpr const char* _type_key = "script.Docs.Doc";
+  TVM_DECLARE_BASE_OBJECT_INFO(DocNode, Object);
+};
+
+class Doc : public ObjectRef {
+ public:
+  Doc() : Doc(make_object<DocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(Doc, ObjectRef, DocNode);
+};
+
+// Base Expr Doc
+
+class ExprDocNode : public DocNode {
+ public:
+  static constexpr const char* _type_key = "script.Docs.BaseExprDoc";
+  TVM_DECLARE_BASE_OBJECT_INFO(ExprDocNode, DocNode);
+};
+
+class ExprDoc : public Doc {
+ public:
+  static ExprDoc TIRBuilder();
+  static ExprDoc None();
+
+  template <typename... AttrType>
+  static ExprDoc TIRBuilderAttribute(AttrType&&... name);
+
+  ExprDoc AccessAttr(String attr);
+
+  template <typename... ArgType>
+  ExprDoc CallWith(ArgType&&... args);
+
+  template <typename... IndexType>
+  ExprDoc IndexWith(IndexType&&... args);
+
+  ExprDoc() : ExprDoc(make_object<ExprDocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(ExprDoc, Doc, ExprDocNode);
+};
+
+// Literal Value Doc
+
+class LiteralValueDocNode : public ExprDocNode {
+ public:
+  ObjectRef value;  // Can only be FloatImm/IntImm/StringImm/String
+
+  static constexpr const char* _type_key = "script.Docs.LiteralValueDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(LiteralValueDocNode, ExprDocNode);
+
+ private:
+  ObjectRef value_;
+};
+
+class LiteralValueDoc : public ExprDoc {
+ public:
+  LiteralValueDoc() : LiteralValueDoc(make_object<LiteralValueDocNode>()){};
+  LiteralValueDoc(IntImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(FloatImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(tir::StringImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(String val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(LiteralValueDoc, ExprDoc, LiteralValueDocNode);
+
+ protected:
+  explicit LiteralValueDoc(ObjectRef value) {
+    auto node = make_object<LiteralValueDocNode>();
+    node->value = std::move(value);
+    data_ = std::move(node);
+  }
+};
+
+// Special Constant Doc
+
+class ConstDocNode : public ExprDocNode {
+ public:
+  enum class ConstKind { TIRBuilder, RelaxBuilder, None };
+
+  ConstKind kind;
+
+  ConstDocNode(ConstKind kind) : kind(kind) {}
+
+  static constexpr const char* _type_key = "script.Docs.ConstDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(ConstDocNode, ExprDocNode);
+};
+
+class ConstDoc : public ExprDoc {
+ public:
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(ConstDoc, ExprDoc, ConstDocNode);
+};
+
+// Identifier Doc
+
+class IdentifierDocNode : public ExprDocNode {
```

Comment:
To clarify the semantics of `IdentifierDoc`.

For the same variable that appears in 2 different places, do they share the same IdentifierDocNode?